gtsfm/frontend/detector_descriptor/detector_descriptor_base.py

```diff
@@ -43,9 +46,46 @@ def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
             Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
         """
 
-    def create_computation_graph(self, image_graph: Delayed) -> Tuple[Delayed, Delayed]:
+    def filter_by_response(self, keypoints: Keypoints, descriptors: np.ndarray) -> Tuple[Keypoints, np.ndarray]:
+        """Filter features according to their responses.
+
+        Args:
+            keypoints: detected keypoints with length M.
+            descriptors: (M, D) array of descriptors D is the dimension of each descriptor.
+
+        Returns:
+            The top N (<= `self.max_keypoints`) keypoints, and their corresponding desciptors as an (N, D) array, with
+                respect to their responses.
+        """
+        if keypoints.responses is None:
+            return keypoints, descriptors
+
+        # Sort by responses.
+        sort_idxs = np.argsort(-keypoints.responses)
+        sort_idxs = sort_idxs[: self.max_keypoints]
+
+        return keypoints.extract_indices(sort_idxs), descriptors[sort_idxs]
+
+    @staticmethod
+    def filter_by_mask(mask: np.ndarray, keypoints: Keypoints, descriptors: np.ndarray) -> Tuple[Keypoints, np.ndarray]:
+        """Filter features with respect to a binary mask of the image.
+
+        Args:
+            mask: (H, W) array of 0's and 1's corresponding to valid portions of the original image.
+            keypoints: detected keypoints with length M.
+            descriptors: (M, D) array of descriptors D is the dimension of each descriptor.
+
+        Returns:
+            N <= M keypoints, and their corresponding desciptors as an (N, D) array, such that their (rounded)
+                coordinates corresponded to a 1 in the input mask array.
         """
-        Generates the computation graph for detections and their descriptors.
+        rounded_coordinates = np.round(keypoints.coordinates).astype(int)
+        valid_idxs = np.flatnonzero(mask[rounded_coordinates[:, 1], rounded_coordinates[:, 0]] == 1)
+
+        return keypoints.extract_indices(valid_idxs), descriptors[valid_idxs]
```

Comment:
Lets modify the get_top_k method to return the keypoints as well as indices and use it here

Reply:
Also, lets move the mask application fucntion to keypoints class